src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,175 @@
+@[simp] lemma map_sub : X (f - f') = X f - X f' := linear_map.map_sub X f f'
+
+instance : add_comm_group (left_invariant_derivation I G) :=
+{ add_assoc := λ X Y Z, ext $ λ a, add_assoc _ _ _,
+  zero_add := λ X, ext $ λ a, zero_add _,
+  add_zero := λ X, ext $ λ a, add_zero _,
+  add_comm := λ X Y, ext $ λ a, add_comm _ _,
+  add_left_neg := λ X, ext $ λ a, add_left_neg _,
+  ..left_invariant_derivation.has_zero,
+  ..left_invariant_derivation.has_add,
+  ..left_invariant_derivation.has_neg }
+
+@[simp] lemma coe_sub : ⇑(X - Y) = X - Y := by simp only [sub_eq_add_neg, coe_add, coe_neg]
+
+instance : module 𝕜 (left_invariant_derivation I G) :=
+module.of_core $
+{ smul := λ r X, ⟨r • X, λ f g, by { simp only [derivation.Rsmul_apply, algebra.id.smul_eq_mul,
+            mul_eq_mul_left_iff, linear_map.map_smul, left_invariant'], left, refl }⟩,
+  mul_smul := λ r s X, ext $ λ b, mul_smul _ _ _,
+  one_smul := λ X, ext $ λ b, one_smul 𝕜 _,
+  smul_add := λ r X Y, ext $ λ b, smul_add _ _ _,
+  add_smul := λ r s X, ext $ λ b, add_smul _ _ _ }
+
+@[simp] lemma map_smul : X (r • f) = r • X f := linear_map.map_smul X r f
+@[simp] lemma smul_map : (r • X) f = r • (X f) := rfl
+@[simp] lemma leibniz : X (f * f') = f • X f' + f' • X f := X.leibniz' _ _
+@[simp] lemma lift_smul (k : 𝕜) : (↑(k • X) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = k • X := rfl
+
+/-- Evaluation at a point for left invariant derivation. Same thing as for generic global
+derivations.-/
+def eval_at : (left_invariant_derivation I G) →ₗ[𝕜] (point_derivation I g) :=
+{ to_fun := λ X, derivation.eval_at g ↑X,
+  map_add' := λ X Y, rfl,
+  map_smul' := λ k X, rfl }
+
+lemma eval_at_apply : eval_at g X f = (X f) g := rfl
+
+@[simp] lemma eval_at_coe : derivation.eval_at g ↑X = eval_at g X := rfl
+
+lemma left_invariant : (𝒅(𝑳 I g)) (1 : G) (eval_at (1 : G) X) f = eval_at g X f :=
+(X.left_invariant'' f g)
+
+lemma eval_at_mul : eval_at (g * h) X f = (𝒅(𝑳 I g)) h (eval_at h X) f :=
+by rw [←left_invariant, L_mul, fdifferential_comp, apply_fdifferential, linear_map.comp_apply,
+  apply_fdifferential, left_invariant]
+
+lemma comp_L : (X f).comp (𝑳 I g) = X (f.comp (𝑳 I g)) :=
+by ext h; rw [times_cont_mdiff_map.comp_apply, L_apply, ←eval_at_apply, eval_at_mul,
+  apply_fdifferential, eval_at_apply]
+
+instance : has_bracket (left_invariant_derivation I G) (left_invariant_derivation I G) :=
+{ bracket := λ X Y, ⟨⁅(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯), Y⁆, λ f g, begin
+    have hX := left_invariant' g X (Y f), have hY := left_invariant' g Y (X f),
+    rw [apply_fdifferential, derivation.eval_apply] at hX hY ⊢, rw [comp_L] at hX hY,
+    rw [derivation.commutator_apply, smooth_map.coe_sub, pi.sub_apply, coe_lift_eq_coe],
+    rw [coe_lift_eq_coe] at hX hY ⊢, rw [hX, hY], refl end⟩ }
+
+@[simp] lemma commutator_coe_derivation :
+  ⇑⁅X, Y⁆ = (⁅(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯), Y⁆ :
+  derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) := rfl
```

Comment:
Is the type annotation on `X` necessary, given the outer type annotation?

Reply:
If I do not include it I run into the error `maximum class-instance resolution depth has been reached`. Is it a local problem on my machine or will it create the same problem on mathlib?

Reply:
Yes, this is needed because `⁅x, y⁆` is not homogenous